src/java.base/share/classes/java/nio/file/CopyMoveHelper.java

```diff
@@ -119,22 +128,42 @@ static void copyToForeignTarget(Path source, Path target,
             throw new FileAlreadyExistsException(target.toString());
 
         // create directory or copy file
-        if (attrs.isDirectory()) {
+        if (srcAttrs.isDirectory()) {
             Files.createDirectory(target);
         } else {
             try (InputStream in = Files.newInputStream(source)) {
                 Files.copy(in, target);
             }
         }
 
-        // copy basic attributes to target
+        // copy basic and, if supported, POSIX attributes to target
         if (opts.copyAttributes) {
-            BasicFileAttributeView view =
-                Files.getFileAttributeView(target, BasicFileAttributeView.class);
+            BasicFileAttributeView dstView = null;
+            if (srcPosixView != null) {
+                dstView = Files.getFileAttributeView(target,
+                                                     PosixFileAttributeView.class);
+            }
+
+            // target might not support posix even if source does
+            if (dstView == null) {
+                dstView = Files.getFileAttributeView(target,
+                                                     BasicFileAttributeView.class);
+            }
+
             try {
-                view.setTimes(attrs.lastModifiedTime(),
-                              attrs.lastAccessTime(),
-                              attrs.creationTime());
+                dstView.setTimes(srcAttrs.lastModifiedTime(),
+                                 srcAttrs.lastAccessTime(),
+                                 srcAttrs.creationTime());
+
+                if (srcAttrs instanceof PosixFileAttributes srcPosixAttrs &&
+                    dstView instanceof PosixFileAttributeView dstPosixView) {
+                    dstPosixView.setPermissions(srcPosixAttrs.permissions());
+                    try {
+                        dstPosixView.setOwner(srcPosixAttrs.owner());
+                        dstPosixView.setGroup(srcPosixAttrs.group());
+                    } catch (ProviderMismatchException ignore) {
```

Comment:
I don't think setOwner/setGroup can work here.  It may be possible to obtain a UserPrincipalLookupService from the target file system and use it to lookup the string representation of the source user/group but it may not be reliable. It's also likely that these setXXX will fail when the target file is on the default file system. Maybe it would be simpler to limit the copying to just POSIX file permissions.

Reply:
So changed in commit beaf175.